Odysseus emagging and new equipment

TDSSS opened the pull request:
**What does this PR do:**
This PR attempts to flesh out the Odysseus a tad and hopefully make it a bit more useful, both for paramedics and for traitors.

**Odysseus can now be emagged**
You can now emag an Ody like you can a Ripley. Like with the Ripley, this gives the description of the mech a red, obvious message telling people something's off, so don't rely on stealth. An emagged Ody can't take weapons like a Ripley can. Instead, several of its tools are modified to make them a bit more dangerous.
In more details:

1. The sleepers of an Emagged Ody can no longer be left by resisting out of them. Instead, a message is displayed to all that can see the Ody showing someone banging against the inside of the sleeper.
2.  Syringe guns unlock Initropidril and Sarin when their Ody gets emagged, unlocking two potent toxins for use by the pilot. As usual, they are also made available to the sleeper.
3. The rescue clamp, a new piece of equipment this PR is adding, becomes capable of doing harm and being used as a weapon. It has the same damage as a ripley's clamp, though no oxy loss.

**New equipment: Rescue clamp**
The original idea for this PR came from this. Namely, from my utter frustration at trying to rescue people in an Ody. Do you know what stops any rescue effort cold? Firelocks. Rushing to the site of a hull breach or plasma fire to save the 5 dying crew member there? Well you're out of luck, because you need to get out, crowbar it open, climb back into your mech and hope the firelock hasn't closed yet.

But no more!
With the patented hydraulic rescue clamp, paramedics can now reach their patients even in the middle of a wrecked, burning station. The hydraulic rescue clamp opens unpowered doors and firelocks just like any crowbar would, from the comfort of your mech. What's more, it can even force powered airlocks, like the jaws of life. Only bolted doors will impede you.
What's more, the rescue clamp is completely unable to do harm! Unlike the rough, industrial equipment of a Ripley, the rescue clamp's inbuilt safeties prevent them from being used as a weapon, turning brutal punches into gentle boops on the nose. Until they get emagged and safety protocols overriden, that is.

**Misc code changes**
I had to do some minor changes to make this work. Emagging of mechs is now properly handled in the subclasses instead of having an istype check on the mech.dm level which checks if it's a Ripley or not. A new helper function was made to check if a tool can be used to force a door, and a minor change to the try_crowbar function. 

**Feature freeze**
To thaw this, I'd like to cash in #9661, #9650 and #9647

**Changelog:**
:cl:
add: New exosuit equipment: The rescue clamp for the Odysseus lets it open doors, firelocks and even force powered doors
add: Odysseus can now be emagged, enabling some rather dangerous uses
/:cl:

SkeletalElite commented:
So just got an idea
Might be a little out of the scope of this PR but what if a malf AI using viral mech domination on a mech emags it if possible.

Pennwick commented:
Initropidril might be too much. Its a very serious drug. 1/3 chance of a 2 tick stun, 1/3 chance of around 15 tox damage, 1/10 chance of some suffocation, and 1/10 chance of straight up heart attack. Its already a blacklisted synth chem for that reason. It CAN be made by chemists but it requires almost every other drug including Bath Salts to produce. Meaning that a Chemist has to get their hands on: Enzyme, Mugwort, ?????, Morphine, and Histamine among MANY other easier ingredients.

Of course the syringes are blocked by hardsuits and the exosuit probably isn't the most sturdy to help balance it but its still a very dangerous chemical that can be abused.

SamHPurp commented on code/game/mecha/medical/odysseus.dm:
```diff
@@ -44,3 +44,21 @@
+	if(emagged == 0)
+		emagged = 1
+		for(var/obj/item/mecha_parts/mecha_equipment/medical/syringe_gun/I in equipment)	//you get some juicy traitor chems for this...
+			I.add_known_reagent("initropidril", "Initropidril")
```
Jesus Christ, no.

Tayyyyyyy commented:
Emags don't need any more buffs, and especially not this buff.

Quantum-M commented:
> The sleepers of an Emagged Ody can no longer be left by resisting out of them. Instead, a message is displayed to all that can see the Ody showing someone banging against the inside of the sleeper.

Does this mean that however is in the sleeper is stuck indefinitely until someone gets into the mech? This seems a bit too strong, considering that the mech can be DNA locked/ID locked, unless the emag removes that, or you could dump said mech into some isolated spot in space with the victim in it.

Ty-Omaha commented:
Clamps...clamp things not pry things

It should be called a jaws of life not a rescue clamp

variableundefined commented:
Judging from claimed features alone:
I echo Ty's naming concerns and everyone's balance concern.

The only thing I can support is adding in the ability for a ody to open firelock and doors. I am also ok with it opening powered airlock.

taukausanake commented:
The only thing I like is the rescue clamp but make it available to a ripley or firefighter ripley as well. The station's on fire, you need to get into a burning room to make it not on fire anymore, oh look door in your way better step outside of my mech and open- OH LORD THE FLAMES

TDSSS commented:
Hope this addresses some of the concerns raised. In order

- You can now struggle free of an emagged sleeper. It takes quite a bit, and if the mech is piloted your kidnapper has quite a bit of time to pump you full of nasty chems or get you somewhere quiet, though.
- Replaced Initropidril with Neurotoxin2
- Renamed to rescue jaw, jaws of life is already the name of the tool the CE gets. Open for suggestions.

@Tayyyyyyy 
One of the best things about the emag is how versatile it is, how almost every machine you can think of does something neat when emagged. So I disagree, it never hurts to give it more options, IMO. Besides, if you have access to an emag and a mech fabricator you can just make a combat ripley or even just a combat mech, not to mention open lockboxes. The only person this really buffs is paramedics and maybe doctors that convince robotics to make them one too. And neither of those are overpowered jobs.

@taukausanake Making them available to firefighter ripleys wouldn't be too hard, what do others think about it?

KasparoVy commented:
jaws of commit not die

Tayyyyyyy commented:
That's exactly why you shouldn't add more to the emag. It's only 8 tc, and constantly adding more stuff to it is making it more and more powerful for the exact same price. The Odysseus can already help transport people that you've killed or captured. It doesn't need to be buffed for traitors.

Also, the Odysseus can already synthesize any chemical. I don't know why it would have in built deadly chemicals.

Overall the other ideas are great, but the emag functionality needs to be completely removed.

variableundefined commented:
@TDSSS As per above, I can support this PR if it only add in a rescue clamp (jaw?) and refactor ody. I cannot support the emag change at all.

TDSSS commented:
Shame, but alright. Removed emag functionality from the Ody. I left the refactor in that moved the emag function to the ripley itself instead of handling it on the mech level, though. Useful in case we ever want to let other mechs be emagged in the future.

Also made the rescue jaw available to firefighter ripleys. Will be updating the OP in a moment.

variableundefined commented:
Please update your PR description accordingly.

I am labelling this as a minor feature and will only charge one fix token.

TDSSS commented:
@variableundefined All done.

Crazylemon64 commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -1042,7 +1042,7 @@ About the new airlock wires panel:
 					close(1)
 
-	if(istype(I, /obj/item/crowbar/power))
+	if(ispowertool(I)) //jaws of life and rescue claw
```
So if someone is holding the rescue claws mech equipment in their hands, will they be able to use it like the jaws of life?

TDSSS replied:
Shouldn't work, because the try_to_crowbar function won't be called if you do that.

TDSSS replied:
I've tried it in game just to verify and can confirm it doesn't work like that.

Crazylemon64 commented on code/game/mecha/mecha.dm:
```diff
@@ -851,14 +851,8 @@
-	else
-		to_chat(usr, "<span class='warning'>The [src]'s ID slot rejects the card.</span>")
+/obj/mecha/emag_act(mob/user)
+	to_chat(user, "<span class='warning'>The [src]'s ID slot rejects the card.</span>")
```
This code has been moved into `ripley.dm`

Fox-McCloud replied:
no `The` before `[thing]`

Crazylemon64 commented on code/modules/research/designs/mechfabricator_designs.dm:
```diff
@@ -653,6 +653,16 @@
+	name = "Exosuit Medical Equipment (Rescue Jaw)"
+	id = "mech_medical_jaw"
+	build_type = MECHFAB
+	req_tech = list("biotech" = 4, "engineering" = 4)
```
design note: the rescue jaw allows progressing from biotech 4->biotech 5

Fox-McCloud replied:
This should honestly be the same cost and tech levels as the jaws of life--same origin tech too.

TDSSS replied:
You could argue it would be a bit easier to make a mech-sized tool capable of forcing doors than it is to make it handheld, but alright. Changing.

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+			target.visible_message("[chassis] gently boops [target] on the nose, its hydraulics hissing as safety overrides slow a brutal punch down at the last second.", \
+								"[chassis] gently boops [target] on the nose, its hydraulics hissing as safety overrides slow a brutal punch down at the last second.")
+		else
+			step_away(M,chassis)//out of the way, I have people to save!
```
Would it be better if this moved the target to the side, instead of directly away from the oddy? As-is it accomplishes the same thing as a walk afaict.

Crazylemon64 replied:
spacing

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+	if(!action_checks(target))
+		return
+	if(istype(target,/obj))
+		if(!istype(target , /obj/machinery/door))//early return if we're not trying to open a door
```
spacing

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+/obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/action(atom/target)
+	if(!action_checks(target))
+		return
+	if(istype(target,/obj))
```
spacing

Crazylemon64 commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1150,6 +1150,11 @@ var/global/list/common_tools = list(
 	return 0
 
+/proc/ispowertool(O)//used to check if a tool can force powered doors
+	if(istype(O, /obj/item/crowbar/power) || istype(O, /obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/))
```
no trailing `/`

Crazylemon64 commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1150,6 +1150,11 @@ var/global/list/common_tools = list(
 
+/proc/ispowertool(O)//used to check if a tool can force powered doors
+	if(istype(O, /obj/item/crowbar/power) || istype(O, /obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/))
+		return 1
```
`TRUE`

Crazylemon64 commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1150,6 +1150,11 @@ var/global/list/common_tools = list(
+/proc/ispowertool(O)//used to check if a tool can force powered doors
+	if(istype(O, /obj/item/crowbar/power) || istype(O, /obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/))
+		return 1
+	return 0
```
`FALSE`

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+
+/obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/can_attach(obj/mecha/M)
+	if(istype(M, /obj/mecha/medical) || istype(M, /obj/mecha/working/ripley/firefighter))	//Odys or firefighters
+		if(M.equipment.len<M.max_equip)
```
Spacing

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+/obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/can_attach(obj/mecha/M)
+	if(istype(M, /obj/mecha/medical) || istype(M, /obj/mecha/working/ripley/firefighter))	//Odys or firefighters
+		if(M.equipment.len<M.max_equip)
+			return 1
```
`TRUE`

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+	if(istype(M, /obj/mecha/medical) || istype(M, /obj/mecha/working/ripley/firefighter))	//Odys or firefighters
+		if(M.equipment.len<M.max_equip)
+			return 1
+	return 0
```
`FALSE`

Crazylemon64 commented on code/game/mecha/working/ripley.dm:
```diff
@@ -180,4 +180,14 @@
+			drill.equip_cooldown = initial(drill.equip_cooldown)
+
+/obj/mecha/working/ripley/emag_act(mob/user)
+	if(emagged == 0)
```
`if(!emagged)`

Crazylemon64 commented on code/game/mecha/working/ripley.dm:
```diff
@@ -180,4 +180,14 @@
+
+/obj/mecha/working/ripley/emag_act(mob/user)
+	if(emagged == 0)
+		emagged = 1
```
`TRUE`

Crazylemon64 commented on code/game/mecha/working/ripley.dm:
```diff
@@ -180,4 +180,14 @@
+		desc += "</br><span class='danger'>The mech's equipment slots spark dangerously!</span>"
+	else
+		to_chat(user, "<span class='warning'>The [src]'s ID slot rejects the card.</span>")
+	return
```
No need for a trailing return

Crazylemon64 commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,38 @@
+	if(isliving(target))	//interact with living beings
+		var/mob/living/M = target
+		if(chassis.occupant.a_intent == INTENT_HARM)//the patented, medical rescue claw is incapable of doing harm. Worry not.
+			target.visible_message("[chassis] gently boops [target] on the nose, its hydraulics hissing as safety overrides slow a brutal punch down at the last second.", \
```
Would some kind of `notice`-class span be appropriate here?

TDSSS replied:
I could see using notice, but that looks very similar, if not identical, doesn't it?

SamHPurp replied:
You need spans on any goon chat messages. Don't leave them naked.

variableundefined commented:
@TDSSS Can you improve your PR title to something like "Add jaw of life to odysseus" instead of "new equipment" when there's only one equipment?

Also, please outright delete the crossed out section in your PR. Using it in moderation with a few words to make a joke is OK, but if it clutter up the entire PR description, it make it hard for people to get to the point.

Keep the description short, concise, and accurate

TDSSS commented:
> @TDSSS Can you improve your PR title to something like "Add jaw of life to odysseus" instead of "new equipment" when there's only one equipment?
> 
> Also, please outright delete the crossed out section in your PR. Using it in moderation with a few words to make a joke is OK, but if it clutter up the entire PR description, it make it hard for people to get to the point.
> 
> Keep the description concise and accurate

I figured it would help people make sense of the discussion if they read it later, but if that's not a concern, sure.

SamHPurp commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,50 @@
+								"[chassis] gently boops [target] on the nose, its hydraulics hissing as safety overrides slow a brutal punch down at the last second.")
+		else
+			push_aside(chassis, M)//out of the way, I have people to save!
+			occupant_message("You gently push [target] out of the way.")
```
again... around here more too

SamHPurp commented on code/game/mecha/equipment/tools/medical_tools.dm:
```diff
@@ -519,3 +519,50 @@
+			occupant_message("You gently push [target] out of the way.")
+			chassis.visible_message("[chassis] gently pushes [target] out of the way.")
+
+/obj/item/mecha_parts/mecha_equipment/medical/rescue_jaw/proc/push_aside(obj/mecha/M, mob/living/L)
```
feels like a helper for this exists already, but could be mistaken and it only exists in TG.

TDSSS replied:
I looked through unsorted.dm and mobs.dm without finding any. While I don't rule out I could have missed some, I don't think it's likely. Want me to turn it into a helper?

SamHPurp commented on code/game/mecha/mecha.dm:
```diff
@@ -851,14 +851,8 @@
-	else
-		to_chat(usr, "<span class='warning'>The [src]'s ID slot rejects the card.</span>")
+/obj/mecha/emag_act(mob/user)
+	to_chat(user, "<span class='warning'>[src]'s ID slot rejects the card.</span>")
```
Seems unnecessary to have a warning that there is no emag action on this... would rather just have emag do nothing at all personally.

TDSSS replied:
That is not something I added, it always worked like that. I just refactored it slightly.

SamHPurp commented on code/game/mecha/working/ripley.dm:
```diff
@@ -180,4 +180,13 @@
+	if(!emagged)
+		emagged = TRUE
+		to_chat(user, "<span class='notice'>You slide the card through [src]'s ID slot.</span>")
+		playsound(loc, "sparks", 100, 1)
```
vol 100 seems pretty high! Everything else tends to cap out at 75... with most at 60.....

TDSSS replied:
See above, not something I changed, I just moved it.

SamHPurp commented on code/game/mecha/working/ripley.dm:
```diff
@@ -180,4 +180,13 @@
+		emagged = TRUE
+		to_chat(user, "<span class='notice'>You slide the card through [src]'s ID slot.</span>")
+		playsound(loc, "sparks", 100, 1)
+		desc += "</br><span class='danger'>The mech's equipment slots spark dangerously!</span>"
```
As an additive description, trying to describe something we should be physically seeing to fall in line with the rest of the codebase seems goofy. If the mech was actively sparking on a regular basis then fair enough, but seems over the top.

TDSSS replied:
Same here. Old part of the codebase.

TDSSS commented:
I've added the span class as requested. However, this means that it is now inconsistent to how the ripley's hydraulic clamp pushes people.
![untitled](https://user-images.githubusercontent.com/32099540/46819177-55481f80-cd83-11e8-8708-a836583da237.png)
As you can see, the ripley's clamp doesn't use any span class. Want me to change that one too?

variableundefined commented:
If you want, might as well fix that one too. Generally speaking we prefer span class on most to-chat to distinguish them.

TDSSS commented:
Did the ripley clamp thing too, since it's such a little thing. That aside, is there anything stopping this from being merged?

variableundefined commented:
It is pending approval.

TDSSS commented:
So nothing more needed from me, awesome. Thanks!
